### ISSUE TITLE
False positive ``used-before-assignmnet`` for assignment expressions in class initialisers

### ISSUE THREAD

**Issue by DanielNoord** [PyCQA/pylint#5212]
### Bug description

Same with classes and dataclasses:

```python
class Dummy:
    def __init__(self, value):
        self.value = value

dummy = Dummy(value=val if (val := 'something') else 'anything')
print(dummy.value)
```

raises `used-before-assigment`, while 
```py
class Dummy:
    def __init__(self, value):
        self.value = value

value = val if (val := 'something') else 'anything'
dummy = Dummy(value)
print(dummy.value)
```
doesn't. Excepting behaviour - no errors.

_Originally posted by @egormm in https://github.com/PyCQA/pylint/issues/3688#issuecomment-924703842_

### Command used

```shell
pylint
```

### Pylint output

```shell
`used-before-assignment` warning in the first example
```

### Expected behavior

No `used-before-assignment` errors for the first example.

### Pylint version

```shell
pylint 2.11.2-dev0
astroid 2.8.4
Python 3.9.6 (default, Aug  2 2021, 08:23:35) 
[Clang 12.0.5 (clang-1205.0.22.11)]
```
